flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -116,16 +140,54 @@
     /** For caching the last accessed partitioned state. */
     private String lastName;
 
+    private final FunctionDelegationHelper functionDelegationHelper =
+            new FunctionDelegationHelper();
+
+    /** Updated initially on restore and later upon materialization (after FLINK-21356). */
+    private final List<KeyedStateHandle> materialized = new ArrayList<>();
+
+    /** Updated initially on restore and later cleared upon materialization (after FLINK-21356). */
+    private final List<ChangelogStateHandle> restoredNonMaterialized = new ArrayList<>();
+
+    /**
+     * {@link SequenceNumber} denoting last upload range <b>start</b>, inclusive. Updated to {@link
+     * #materializedTo} when {@link #snapshot(long, long, CheckpointStreamFactory,
+     * CheckpointOptions) starting snapshot}. Used to notify {@link #stateChangelogWriter} about
+     * changelog ranges that were confirmed or aborted by JM.
+     */
+    @Nullable private SequenceNumber lastUploadedFrom;
+    /**
+     * {@link SequenceNumber} denoting last upload range <b>end</b>, exclusive. Updated to {@link
+     * org.apache.flink.runtime.state.changelog.StateChangelogWriter#lastAppendedSequenceNumber}
+     * when {@link #snapshot(long, long, CheckpointStreamFactory, CheckpointOptions) starting
+     * snapshot}. Used to notify {@link #stateChangelogWriter} about changelog ranges that were
+     * confirmed or aborted by JM.
+     */
+    @Nullable private SequenceNumber lastUploadedTo;
+    /**
+     * The {@link SequenceNumber} up to which the state is materialized, exclusive. The log should
+     * be truncated accordingly.
+     *
+     * <p>WARN: currently not updated - to be changed in FLINK-21356.
```

Comment:
This PR targets for FLINK-21356 and field `materializedTo` is still not updated, this really confuse me.

Reply:
Without materialization, the backend is still incomplete and my intention was to make it clear what else and how will change. It's not ideal, but leaving `lastUploadedFrom` never updated (or always updated to 0) would have the same or worse effect IMO.
I can remove it though if you'd like.

Reply:
@rkhachatryan I thought it's the statement bring some confusion? Maybe 'to be change AFTER FLINK-21356'(or 'IN' any other FLINK issue) could be better? @Myasuka, is that your concern?

Reply:
I think `materializedTo` would be updated in later PR (not FLINK-21356). If so, `to be changed in FLINK-21356` seems not correct.

Reply:
Oh, sure, it should be FLINK-21357 not FLINK-21356 🙃
I'll update the comment if you're fine with that.
Sorry for the confusion.